### PR TITLE
Fix path navigation in FileBrowser and DirBrowser

### LINKED ISSUE #547: Navigating filesystem in FileBrowser results in incorrect paths
**Newsboat version (copy the output of `newsboat -v` or the first line of `git show`)**: newsboat r2.15-235-g3a1e0
System: Linux 4.19.0-5-amd64 (x86_64)
Compiler: g++ 8.3.0
ncurses: ncurses 6.1.20181013 (compiled with 6.1)
libcurl: libcurl/7.64.0 GnuTLS/3.6.7 zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.36.0 librtmp/2.3 (compiled with 7.64.0)
SQLite: 3.27.2 (compiled with 3.27.2)
libxml2: compiled with 2.9.4

Steps to reproduce the issue:

1. Open Newsboat, navigate to any article, press `s` to save it. "FileBrowser" dialog is opened.

2. Press `Tab` to move focus from the input field to files list.

3. The cursor is on `..`. Press `Enter` to navigate up by the file hierarchy. In my place, I started in `/home/minoru`, so I'm now in `/home/`, and the path to file (at the bottom of the dialog) says `/home/article_title.txt`, which seems correct.

4. Use `Down` to scroll to some directory, and press `Enter`. I picked `minoru`, and thus moved to `/home/minoru`. However, the path have been updated to `/home/minoru//home/article_title.txt`.

5. Every time I change directories, the new dirpath is prepended to the current path.

**Other info you think is relevant**:

The same currently exists in DirBrowser (#531), because it copies a lot of FileBrowser's code.

### PR DESCRIPTION
**Pull request by dennisschagt**
Fixes #547.

### CODE REVIEW

**Comment by dennisschagt** [comment]
> This indeed fixes #547, but also adds two new bugs (although minor compared to #547):
> 
> 1. if I start in /home/minoru, the title of FileDirBrowser says "newsboat r2.18-45-gebee - Save File - **/home/minoru**". But when I navigate one level up, the title says "newsboat r2.18-45-gebee - Save File - **..**" instead of "newsboat r2.18-45-gebee - Save File - **/home**".

It looks like this behaviour already was present in the latest release (2.18) and on master.
I think it is better if I create a separate issue/PR for that.

> 2. If I navigate to the filesystem root, "Filename" input says "//ArticleName.txt" -- with two slashes instead of expected one slash.

Fixed

File: src/dirbrowserformaction.cpp [hunk 1]
```diff
@@ -106,7 +106,9 @@ void DirBrowserFormAction::process_operation(Operation op,
 						status);
 					f->set("listpos", "0");
 					std::string fn = utils::getcwd();
-					fn.append(NEWSBEUTER_PATH_SEP);
+					if (fn.back() != '/') {
```
Comment (Minoru) [inline src/dirbrowserformaction.cpp]: I think it'd be better to turn `NEWSBEUTER_PATH_SEP` into a char (it's currently a string literal), and use it here. I don't think we'll ever turn that constant into an actual setting, and it's bound to disappear once we move to Rust, but in the meantime, it still feels wrong to hard-code the char here while still using the constant on the next line.

**Comment by Minoru** [comment]
> It looks like this behaviour already was present in the latest release (2.18) and on master.

Yes, sorry — my bad. I'll move that into a separate issue.

File: src/dirbrowserformaction.cpp [hunk 2]
```diff
@@ -106,13 +106,13 @@ void DirBrowserFormAction::process_operation(Operation op,
 						status);
 					f->set("listpos", "0");
 					std::string fn = utils::getcwd();
-					if (fn.back() != '/') {
-						fn.append(NEWSBEUTER_PATH_SEP);
+					if (fn.back() != NEWSBEUTER_PATH_SEP) {
+						fn.append(1, NEWSBEUTER_PATH_SEP);
```
Comment (Minoru) [inline src/dirbrowserformaction.cpp]: Wouldn't `fn.push_back(NEWSBEUTER_PATH_SEP)` be clearer?

File: src/dirbrowserformaction.cpp (hunk 1 above)
Comment (dennisschagt) [reply src/dirbrowserformaction.cpp]: Makes sense, done.

I have replaced most of the hardcoded uses of `'/'` except where it is used for something else than paths (e.g. urls or tags).

File: src/pbcontroller.cpp [hunk 3]
```diff
@@ -177,9 +177,9 @@ PbController::PbController()
 	}
 
 	config_file =
-		config_dir + std::string(NEWSBEUTER_PATH_SEP) + config_file;
-	queue_file = config_dir + std::string(NEWSBEUTER_PATH_SEP) + queue_file;
-	lock_file = config_dir + std::string(NEWSBEUTER_PATH_SEP) + lock_file;
+		config_dir + NEWSBEUTER_PATH_SEP + config_file;
```
Comment (Minoru) [inline src/pbcontroller.cpp]: Looks like this statement is short enough to fit into a single line now, so please remove the newline after `=`.

File: src/dirbrowserformaction.cpp (hunk 2 above)
Comment (dennisschagt) [reply src/dirbrowserformaction.cpp]: Yes, definitely.
I changed it for all occurrences of `std::string::append(1, ..)`.

By the way, let me know if ever you think I made some changes which you rather see in a separate pull request.

Comment (Minoru) [reply src/dirbrowserformaction.cpp]: > By the way, let me know if ever you think I made some changes which you rather see in a separate pull request.

Sure thing! These `append(1, x) -> push_back(x)` changes are too small to warrant a small PR though, it's fine to bundle them here.

**Comment by Minoru** [comment]
Thank you for the PR, @dennisschagt!

**Pull request merged**